### DISCUSSION TITLE
Not getting the latest version

### DISCUSSION THREAD

**Discussion by hslavchev** [dvershinin/lastversion#72] [General]
Found a mismatch:
```
lastversion https://github.com/mtougeron/k8s-aws-ebs-tagger/releases
1.2.2
```
If you go to the URL above you can see the latest release is [v0.0.5](https://github.com/mtougeron/k8s-aws-ebs-tagger/releases/tag/v0.0.5)

Also in projects like: https://github.com/kubernetes/autoscaler/releases where there are multiple different types of releases (e.g. there's a helm chart version as well as an app version) would be nice to be able to specify which one you want to get.

Oh and one more suggestion: Artifacthub.io could be added as a repo to check. This could help with the above issue since it has both Application version and Chart version (e.g. https://artifacthub.io/packages/helm/cluster-autoscaler/cluster-autoscaler)

**Comment by dvershinin** [comment]
I think you're looking for `--only` switch. This works in latest `lastversion` release:

```console
lastversion --only chart https://github.com/mtougeron/k8s-aws-ebs-tagger/releases
1.2.2
```

```console
lastversion --only v https://github.com/mtougeron/k8s-aws-ebs-tagger/releases
0.0.5
```

However, to find app version (negating "chart" in tags) like in your second example, wasn't possible, so I've added `--only-not` in `lastversion` master:

```console
lastversion --only-not chart https://github.com/kubernetes/autoscaler/releases
1.23.0
```

The `--only-not` is really just using/negating `--only`, so they are self exclusive...

Interesting about Artifacthub.io, but how do you see invoking lastversion to get either chart or app version?

If Artifacthub.io has an API of sorts "give me Artifachub.io helm chart for this Github repo", then we can make using existing "branch" notation of `lastversion`, e.g.

    lastversion kubernetes/autoscaler:chart

**Reply by hslavchev** [reply]
Thanks! I'm actually using lastversion as a python module, and looks like I missed the `only` param.

The `--only-not` could be just `--exclude` maybe.

Regarding Artifacthub.io, it does have an API. I imagine this could be used to get both chart and app versions : https://artifacthub.io/docs/api/#/Packages

**Comment by dvershinin** [comment]
Alright, I've added the `exclude` parameter in the latest release.

Please see https://github.com/dvershinin/lastversion/issues/50, as there's already `at=helm_chart`.
This handles a case when a GitHub project hosts many charts and none corresponding to a GitHub tag or release.
I think how to combine this with artifactshub, so they can be queried with lastversion in a consistent manner.

My current suggestion is to add further logic to it, like the following:

#### Current function

```console
# for full URLs with at=helm_chart, will append '/Chart.yml' and use its data
lastversion.latest('https://github.com/bitnami/charts/blob/master/bitnami/apache', at='helm_chart')
```

### To be added function

Using [API](https://artifacthub.io/docs/api/#/Packages/getHelmPackageDetails):

```console
# only for full URLs at artifactshub: at=helm_chart is implied based on '/packages/helm/', and we fetch data from that API
lastversion.latest('https://artifacthub.io/packages/helm/cluster-autoscaler/cluster-autoscaler')

# for short URIs of only 2 components (one slash), use artifacthub only when `at=helm_chart` is explicitly given:
lastversion.latest('cluster-autoscaler/cluster-autoscaler', at='helm_chart')
```

Any thoughts?

**Reply by hslavchev** [reply]
That could work, however, when looking at helm charts you have two versions: 1) the chart version and 2) the app version. For my specific use case I'm really only looking for the app version and `at=helm_chart` provides the chart version.

My original thought was that there would be an `at=artifacthub` (which may also be more intuitive) which could provide the desired version (app/chart), maybe controlled with a switch. Not sure about the best way to do this to be honest.

The good thing about ArtifactHub is that it's sort of "official" and is even linked to from helm's home page: (https://helm.sh/)
